demo/sieve/src/main.rs

```diff
@@ -0,0 +1,149 @@
+//! Sieve of Eratosthenes
+//!
+//! A sieve finds prime numbers by first assuming all candidates are prime,
+//! then progressively using small primes to mark their multiples composite.
+//!
+//! Note that a given prime p only needs to start marking its multiples >=p²,
+//! as anything less will already have been marked by an even smaller prime
+//! factor.  This also means that sieving up to N only requires primes up to
+//! sqrt(N) - a useful fact to prepare for parallel sieving.
+//!
+//! Here we have three forms of the sieve:
+//!
+//! - `sieve_serial`: direct iteration in serial
+//! - `sieve_chunks`: chunked iteration, still serial
+//! - `sieve_parallel`: chunked iteration in parallel
+//!
+//! The performance improvement of `sieve_chunks` is largely due to cache
+//! locality.  Since the sieve has to sweep repeatedly over the same memory,
+//! it's helpful that the chunk size can fit entirely in the CPU cache.  Then
+//! `sieve_parallel` also benefits from this as long as there's cache room for
+//! multiple chunks, for the separate jobs in each thread.
+
+extern crate itertools;
+extern crate rayon;
+extern crate time;
+
+use itertools::StrideMut;
+use rayon::prelude::*;
+
+const CHUNK_SIZE: usize = 100_000;
+
+const MAX: usize = 1_000_000_000;
+
+// Number of Primes < 10^n
+// https://oeis.org/A006880
+const NUM_PRIMES: usize = 50_847_534;
+
+// For all of these sieves, sieve[i]==true -> 2*i+1 is prime
+
+/// Sieve odd integers for primes < max.
+fn sieve_serial(max: usize) -> Vec<bool> {
+    let mut sieve = vec![true; max / 2];
+    sieve[0] = false; // 1 is not prime
+    for i in 1 .. {
```

Comment:
Yeah, you could do a square root and determine that break point ahead of time, but I suspect it won't matter much to performance.  Those branches should be very easy for the CPU branch predictor, and this is not a hotspot.  The code in `clear_stride` has more impact.  But please feel free to try some optimizations yourself!